File: SimilarNet/SimilarNet/HunkDiff.cs
namespace SimilarNet;

public enum DiffType
{
    Unmodified = 0,
    Deleted = 1,
    Inserted = 2,
    Modified = 3,
}

public class HunkDiff
{
    public DiffType DiffType;
    public Unmodified? Unmodified;
    public Deleted? Deleted;
    public Inserted? Insert;
    public Modified? Modified;
}

public struct Unmodified
{
    public Range OldRange;
    public Range NewRange;
}

public struct Deleted
{
    public Range OldRange;
    public int NewStartLine;
}

public struct Inserted
{
    public int OldStartLine;
    public Range NewRange;
}

public struct Modified
{
    public Range OldRange;
    public Range NewRange;
}

File: similar-rs/src/hunk_diff.rs
use std::ops::Range;

#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub enum DiffType {
    Unmodified = 0,
    Deleted = 1,
    Inserted = 2,
    Modified = 3,
}

#[repr(C)]
#[derive(Debug, Clone)]
pub struct HunkDiff {
    pub diff_type: DiffType,
    pub unmodified: *const Unmodified,
    pub deleted: *const Deleted,
    pub insert: *const Inserted,
    pub modified: *const Modified,
}

#[repr(C)]
#[derive(Debug, Clone)]
pub struct HunkDiffArray {
    pub count: i32,
    pub diffs: *const HunkDiff,
}

#[repr(C)]
#[derive(Debug, Clone)]
pub struct Unmodified {
    pub old_range: Range<i32>,
    pub new_range: Range<i32>,
}

#[repr(C)]
#[derive(Debug, Clone)]
pub struct Deleted {
    pub old_range: Range<i32>,
    pub new_start_line: i32,
}

#[repr(C)]
#[derive(Debug, Clone)]
pub struct Inserted {
    pub old_start_line: i32,
    pub new_range: Range<i32>,
}

#[repr(C)]
#[derive(Debug, Clone)]
pub struct Modified {
    pub old_range: Range<i32>,
    pub new_range: Range<i32>,
}
